src/gretel_trainer/benchmark/session.py

```diff
@@ -32,136 +29,75 @@
-
-        self.config.working_dir.mkdir(exist_ok=True)
-        self.datasets = [
-            _standardize_dataset(dataset, self.config.working_dir)
-            for dataset in datasets
-        ]
+        self._jobs = jobs
+        self._config = config or BenchmarkConfig()
+
+        _validate_jobs(self._config, jobs)
+
         self._gretel_executors: dict[RunKey, Executor] = {}
         self._custom_executors: dict[RunKey, Executor] = {}
-        self.trainer_project_names: dict[str, str] = {}
-        self.thread_pool = ThreadPoolExecutor(5)
-        self.futures: dict[FutureKeyT, Future] = {}
+        self._trainer_project_names: dict[str, str] = {}
+        self._thread_pool = ThreadPoolExecutor(5)
+        self._futures: dict[FutureKeyT, Future] = {}
 
         self._report_scores = {
             score_name: ALL_REPORT_SCORES[score_name]
-            for score_name in ["SQS"] + self.config.additional_report_scores
+            for score_name in ["SQS"] + self._config.additional_report_scores
         }
 
     @property
     def executors(self) -> dict[RunKey, Executor]:
         return self._gretel_executors | self._custom_executors
 
     def _make_project(self) -> Project:
-        display_name = self.config.project_display_name
-        if self.config.trainer:
+        display_name = self._config.project_display_name
+        if self._config.trainer:
             display_name = f"{display_name}-evaluate"
 
         return create_project(display_name=display_name)
 
-    def prepare(self) -> Comparison:
+    def prepare(self) -> Session:
         self._project = self._make_project()
 
         _trainer_project_index = 0
-        for dataset in self.datasets:
-            # TODO: avoid upload for public datasets, when possible
-            artifact_key = _upload_dataset_to_project(
-                dataset.data_source,
-                self._project,
-                self.config.trainer,
-            )
 
-            for model in self.gretel_models:
-                self._setup_gretel_run(
-                    dataset, model, artifact_key, _trainer_project_index
+        data_source_map = {}
+        for job in self._jobs:
+            if (artifact_key := data_source_map.get(job.dataset.data_source)) is None:
+                # TODO: avoid upload for public datasets, when possible
+                artifact_key = _upload_dataset_to_project(
+                    job.dataset.data_source,
+                    self._project,
+                    self._config.trainer,
                 )
+                data_source_map[job.dataset.data_source] = artifact_key
+
+            if is_gretel_model(job):
+                self._setup_gretel_run(job, artifact_key, _trainer_project_index)
                 _trainer_project_index += 1
 
-            for model in self.custom_models:
-                self._setup_custom_run(dataset, model, artifact_key)
+            else:
+                assert is_custom_model(job)
```

Comment:
This assert is needed to make the next line happy. I initially had `elif`, but I think this is better, as there are no other cases.
There are some discussions on type narrowing in the negative case [here](https://github.com/python/typing/discussions/1013), but there is nothing yet that could be used.

Reply:
Interesting read in that discussion, thanks for linking to it! I think I would've landed on using `elif` here personally, but I don't feel too strongly about it. Obviously we shouldn't ever hit this, but I suppose if this assertion ever does explode at runtime we'll have 1) an extremely unusual/unexpected edge case on our hands, and 2) hopefully a better lead on resolving the situation than whatever the error message would end up being otherwise?

Is it worth a comment in the code?

Reply:
Good call. I think between adding a comment and switching it to `elif` + `else` with exception, which acts as a comment the latter sounds like a better option, so I changed it to that.